### DISCUSSION TITLE
BlenderBot2 ValueError: invalid server in colab

### DISCUSSION THREAD

**Discussion by Jawad1347** [facebookresearch/ParlAI#3896] [Q&A] [answered]
Hello, when I try to follow tutorial ParlAI colab, it works fine until I use blender bot 2. (It works fine with blender bot 1, at least for chatting, which I have checked so far). The following error is encountered whenever I try running blenderbot2.

### ERROR
Successfully installed huggingface-hub-0.0.12 sacremoses-0.0.45 transformers-4.9.1
13:37:53 | Overriding opt["model_file"] to /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/blenderbot2_400M/model (previously: /checkpoint/kshuster/projects/knowledge_bot/kbot_memfix_sweep25_Fri_Jul__9/338/model.oss)
13:37:53 | Using CUDA
13:37:53 | loading dictionary from /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/blenderbot2_400M/model.dict
13:37:54 | num words = 50264
13:37:54 | Downloading https://dl.fbaipublicfiles.com/fairseq/gpt2_bpe/vocab.bpe to /usr/local/lib/python3.7/dist-packages/data/gpt2/vocab.bpe
Downloading vocab.bpe: 0.00B [00:01, ?B/s]13:37:55 | Downloading https://dl.fbaipublicfiles.com/fairseq/gpt2_bpe/encoder.json to /usr/local/lib/python3.7/dist-packages/data/gpt2/encoder.json

Downloading encoder.json: 0.00B [00:01, ?B/s]
13:37:58 | BlenderBot2Fid: full interactive mode on.
Downloading: 100%
232k/232k [00:05<00:00, 43.4kB/s]

Downloading: 100%
466k/466k [00:03<00:00, 134kB/s]

Downloading: 100%
28.0/28.0 [00:00<00:00, 410B/s]

Downloading: 100%
570/570 [00:00<00:00, 9.45kB/s]

>13:38:20 | building data: /usr/local/lib/python3.7/dist-packages/data/models/hallucination/bart_rag_token/model.tgz
13:38:20 | Downloading http://parl.ai/downloads/_models/hallucination/bart_rag_token/model.tgz to 
/usr/local/lib/python3.7/dist-packages/data/models/hallucination/bart_rag_token/model.tgz
Downloading model.tgz: 100%|██████████| 950M/950M [01:32<00:00, 10.3MB/s]
13:40:15 | Creating the search engine retriever.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-264b09131cef> in <module>()
      9     # Model files that begin with "zoo:" are special files distributed by the ParlAI team.
     10     # They'll be automatically downloaded when you ask to use them.
---> 11     model_file='zoo:blenderbot2/blenderbot2_400M/model'
     12     # model_file= "zoo:blenderbot2/blenderbot2_3B/model"
     13     # model_file='zoo:blender/blender_400Mdistill/model'

16 frames
/usr/local/lib/python3.7/dist-packages/parlai/agents/rag/retrieve_api.py in _validate_server(self, address)
     97     def _validate_server(self, address):
     98         if not address:
---> 99             raise ValueError('Must provide a valid server for search')
    100         if address.startswith('http://') or address.startswith('https://'):
    101             return address

ValueError: Must provide a valid server for search

### Code

!pip install -q parlai
!pip install -q subword_nmt
!pip install transformers
import transformers
from parlai.scripts.interactive import Interactive

Interactive.main(
    # the model_file is a filename path pointing to a particular model dump.
    # Model files that begin with "zoo:" are special files distributed by the ParlAI team.
    # They'll be automatically downloaded when you ask to use them.
    model_file='zoo:blenderbot2/blenderbot2_400M/model'
)

### one extra question
What is the 950M thing that gets downloaded with the model? Model is 2.42GB. That thing that says "bart_rag_token/model.tgz"

**Comment by mojtaba-komeili** [comment, marked as answer]
Blenderbot 2 needs to connect to a search server for accessing web.
Do you have a running search server? And if you do, do you pass its correct address to the model?

**Reply by Jawad1347** [reply]
thank you for answering:

I don't know what a search server is. Also have few other questions.

Why blenderbot2 downloads so many other models? Like
950MB BART from http://parl.ai/downloads/_models/hallucination/bart_rag_token/model.tgz
875MB BERT base from  https://dl.fbaipublicfiles.com/dpr/checkpoint/retriver/multiset/hf_bert_base.cp 
750MB query generator and decoder from
 http://parl.ai/downloads/_models/blenderbot2/query_generator/model.tgz 
 http://parl.ai/downloads/_models/blenderbot2/memory_decoder/model.tgz
and then there is another BART of about 3.70GB

Why does fully trained model like blender require these thing?

**Reply by klshuster** [reply]
please search the issues tab for information about search servers. blenderbot2 is an amalgamation of several modules:

- the memory retriever is a FiD-RAG retriever, so the bart_rag_token model is required to download to extract the DPR model from within (more information about FiD-RAG can be found [in this project](https://parl.ai/projects/hallucination/)
- BERT-base is required to get the document encoder for the DPR model in the memory retriever
- query generator/memory decoder are what generate the search queries and write the memories, as described in the blog post

**Comment by Jawad1347** [comment]
Another question. In https://parl.ai/projects/blenderbot2/ blog it says that it can pull information from internet. But it did not know who was current president of USA and said it was trump.

![big_convo_final](https://user-images.githubusercontent.com/60029101/128202464-b9f91635-4366-41a3-9d8c-4865418d97ac.jpg)

**Reply by klshuster** [reply]
I am not sure what you mean, I don't see Trump mentioned in the linked picture

**Reply by Jawad1347** [reply]
Never mind I have gotten the answer (through trial and error).
However I meant, that picture says that it looks for information online, while I was unable to determine whether the model was taking information from outside because it was returning the answer to the question "who is current president of USA"? as "Donald Trump". Even when it can find information outside, (with valid server), it still first looks for it inside itself (hence sometimes answer to questions like "what year this is" is given as "2017".)
